documentation/tutorials/get-started-with-ash-and-phoenix.md

```diff
@@ -0,0 +1,481 @@
+    # Add a string type attribute called `:title`
+    attribute :title, :string do
+      # We don't want the title to ever be `nil`
+      allow_nil? false
+    end
+
+    # Add a string type attribute called `:content`
+    # If not `allow_nil?` specified content can be `nil`
+    attribute :content, :string
+  end
+end
+```
+
+### Migrate Database
+
+We have specified the resource, attributes and actions in Ash. But we have yet to create them in our data layer (in our case postgres).
+
+We created our database earlier, but now we need to populate it. We do this by generating and performing a migration.
+
+We can use a generator to produce a migration for us. Ash can deduce what needs to go into the migration and do the hard work for us, just use the command below:
+
+```bash
+$ mix ash_postgres.generate_migrations --name initial_migration
+
+# ... don't worry about other files it creates
+
+Generating Migrations:
+* creating priv/repo/migrations/20230208045101_initial_migration.exs
+```
+
+Inside the migration file we find:
+
+```elixir
+defmodule MyAshPhoenixApp.Repo.Migrations.InitialMigration do
+  use Ecto.Migration
+
+  # this function runs when migrating forward
+  def up do
+    # creates the `:post` table
+    create table(:posts, primary_key: false) do
+      # adds primary key attribute `:id` of type `:uuid`
+      # nil values are not allowed
+      add :id, :uuid, null: false, default: fragment("uuid_generate_v4()"), primary_key: true
+
+      # adds attribute `:title` of type `:text`, nil values are not allowed
+      add :title, :text, null: false
+      # adds attribute `:content` of type `:text`, nil values are allowed
+      add :content, :text
+    end
+  end
+
+  # this is the function that runs if you want to rollback this migration.
+  def down do
+    # deletes the `:post` table
+    drop table(:posts)
+  end
+end
+```
+
+We can run this function which will run this up function and perform these operations on the postgres database. We Run this command:
+
+We can run the `up/0` function which will perform the desired operations on the postgres database. We do this with this command:
+
+```bash
+$ mix ash_postgres.migrate
+```
+
+### Interacting with your resource
+
+All interaction with your resource attributes **ALWAYS** occur through an **action**. In our resource we are using the default actions for `:create, :read, :update, :destroy`. Create, update and destroy actions **ALWAYS** take a changeset. Ash changesets are conceptually similar to [Ecto changesets](https://hexdocs.pm/ecto/Ecto.Changeset.html). They're data structures which represent an intended change to an Ash resource.
+
+Let's write a test to show how to interact with our resource.
+
+```elixir
+defmodule MyAshPhoenixApp.Blog.PostTest do
+  use MyAshPhoenixApp.DataCase, async: true
+
+  test "testing blog post actions" do
+    ### CREATE ACTION - create new blog post ###
+    # We create an Ash changeset we intent to use to create an Ash resource
```

Comment:
We are calling the action, but through the api. When you create a `code_interface` this is what is generated in the background, the code interface is a shortcut. I didn't want to have to explain the code interface, and also want whoevers reading to know about changesets as they're important when it comes to defining actions.

Reply:
so if you have a `:create` change called `:custom_create_change` it could be written like this:

```elixir
create :custom_create_change do
  ...
end
```

This can be called like so:
```elixir
Ash.Changeset.for_create( # we're using for_create because the action is of create type
  MyAshPhoenixApp.Blog.Post,
  :custom_create_change, # we're using the _name_ of the specific create action
  %{title: "hello world"}
)
|> MyAshPhoenixApp.Blog.create!()
```

Reply:
anyway after the discussion we had on slack I think I'm going to adapt this part and include an explanation about the code interface. Then incorporate it into a phoenix live view example